### PR TITLE
[luci] Add RemoveRedundantQuantizePass

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This commit adds new pass for removing redundant quantize operations in luci.

for issue: https://github.com/Samsung/ONE/issues/8526

first step of #8530

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

File: compiler/luci/pass/src/RemoveRedundantQuantizePass.test.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  int count = 0;
+  for (auto node : loco::active_nodes(loco::output_nodes(g.g())))
+  {
+    if (dynamic_cast<luci::CircleQuantize *>(node))
+    {
+      count++;
+    }
+  }
+
+  ASSERT_EQ(0, count);
```
Comment (binarman) [inline compiler/luci/pass/src/RemoveRedundantQuantizePass.test.cpp:166]: This assert fails, actual count is 1

File: compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+bool remove_redundant_quantize(luci::CircleQuantize *node)
+{
+  auto pred_node = loco::must_cast<luci::CircleNode *>(node->input());
+
+  if (pred_node == nullptr)
+    return false;
+
+  if (node->quantparam() == nullptr or pred_node->quantparam() == nullptr)
+    return false;
```
[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp:58]: ping @jinevening

File: compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace
+{
+
+bool remove_redundant_quantize(luci::CircleQuantize *node)
+{
+  auto pred_node = loco::must_cast<luci::CircleNode *>(node->input());
+
+  if (pred_node == nullptr)
+    return false;
```
Comment (jinevening) [inline compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp:55]: `loco::must_cast` ensures that `pred_node` is not nullptr, so this check is redundant.

File: compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jinevening) [reply compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp:58]: This pass is for quantized (or partially quantized) models. And as mentioned offline, `node` would have qparam as it is Quantize Op.

File: compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+} // namespace
+
+namespace luci
+{
+
+bool RemoveRedundantQuantizePass::run(loco::Graph *g)
+{
+  bool changed = false;
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
```
[the remark this answers is not in the document]
Comment (jinevening) [reply compiler/luci/pass/src/RemoveRedundantQuantizePass.cpp:96]: I guess that the final result would be the same if the pass is run in the `Saturate` mode, but using `loco::postorder_traversal` would be better to reduce iterations.

File: compiler/luci/pass/src/RemoveRedundantQuantizePass.test.cpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      auto quantize_param = std::make_unique<luci::CircleQuantParam>();
+      quantize_param->scale = {0.5};
+      quantize_param->zerop = {0};
+      input()->quantparam(std::move(quantize_param));
+    }
+
+    _first_quantize->input(input());
+    _second_quantize->input(_first_quantize);
+
+    output()->from(_second_quantize);
```
Comment (jinevening) [inline compiler/luci/pass/src/RemoveRedundantQuantizePass.test.cpp:113]: Maybe?

```suggestion
    output()->from(_first_quantize);
```

**Review by jinevening — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by binarman — APPROVED** [review]
LGTM

**Pull request merged**